### ISSUE TITLE
Sourcemaps doesn't work

### ISSUE THREAD

**Issue by TrySound** [w0rm/gulp-postcss#18]
I compile css with this

```
	return gulp.src(bower(['**/*.css']).concat([
			'src/css/plugins/[^_]*.css',
			'src/css/globals/[^_]*.css',
			'src/css/shared/[^_]*.css',
			'src/css/components/[^_]*.css',
			'src/css/sections/[^_]*.css'
		]))
		.pipe($.sourcemaps.init())
		.pipe($.concat('bundle.css'))
		.pipe($.plumber())
		.pipe($.postcss([
			require('autoprefixer-core')()
		]))
		.pipe(gulp.dest('dist/assets/css'))
		.pipe($.rename({suffix: '.min'}))
		.pipe($.plumber())
		.pipe($.postcss([
			require('csswring')({
				removeAllComments: true
			})
		]))
		.pipe($.sourcemaps.write('.'))
		.pipe(gulp.dest('dist/assets/css'));
```

And chrome gave me `bundle.css` in sources with `bundle.min.css` name.
When I commented both postcss pipes, it worked.

**Comment by w0rm** [comment]
@TrySound its strange. Can you please try to reduce it to a minimum test case and provide it including your package.json?

**Comment by TrySound** [comment]
@w0rm https://github.com/TrySound/gulp-postcss-sourcemaps-test

**Comment by w0rm** [comment]
@TrySound thanks it was helpful! Please don't delete your repo. 

So currently sourcemaps pass these three steps:

```
CONCAT
 { version: 3,
  sources: [ 'a.css', 'b.css', 'c.css' ],
  names: [],
  mappings: 'AAAA;AACA;AACA;AACA;ACHA;AACA;AACA;AACA;ACHA;AACA;AACA;AACA',
  file: 'result-postcss.css',
  sourcesContent: 
   [ '.a {\n\ttext-align: center;\n}\n',
     '.b {\n\ttext-align: center;\n}\n',
     '.c {\n\ttext-align: center;\n}\n' ] }

POSTCSS
 { version: 3,
  sources: [ 'src/result-postcss.css' ],
  names: [],
  mappings: 'AAAA;CACC,oBAAmB;EACnB;;AAED;CACC,oBAAmB;EACnB;;AAED;CACC,oBAAmB;EACnB',
  file: 'src/result-postcss.css',
  sourcesContent: [ '.a {\n\ttext-align: center;\n}\n\n.b {\n\ttext-align: center;\n}\n\n.c {\n\ttext-align: center;\n}\n' ] }

RESULT
 { version: 3,
  sources: [ 'src/result-postcss.css' ],
  names: [],
  mappings: 'AAAA;CACC,oBAAmB;EACnB;;AAED;CACC,oBAAmB;EACnB;;AAED;CACC,oBAAmB;EACnB',
  file: 'src/result-postcss.css',
  sourcesContent: [ '.a {\n\ttext-align: center;\n}\n\n.b {\n\ttext-align: center;\n}\n\n.c {\n\ttext-align: center;\n}\n' ] }
```

Problem that I see now is mismatch of the file attribute in sourcemaps:  `result-postcss.css` after concat step and `src/result-postcss.css` after postcss step. 

Currently gulp-postcss sets PostCSS `from` option to `file.path` which is absolute path of the gulp file.
For result-poscss.css it will be `/Users/username/path/to/gulp-postcss-sourcemaps-test/src/result-postcss.css` what is then resolved to `src/result-postcss.css` in postcss sourcemap.

In this case I believe that gulp-concat outputs incorrect data, saying that file is in 'src/result-postcss.css' directory, but sourcemap links to 'result-postcss.css'.

cc @floridoo @ai

**Comment by w0rm** [comment]
@TrySound I found ugly hack to fix your problem, use `gulp-rename` and put `.pipe(rename({dirname: '..'}))` after concat and before postcss. So you will have correct output

```
POSTCSS
 { version: 3,
  sources: [ 'result-postcss.css' ],
  names: [],
  mappings: 'AAAA,GACC,kBAAmB,AAGpB,CAFC,EAGA,kBAAmB,AAGpB,CAFC,EAGA,kBAAmB,CACnB',
  file: 'result-postcss.css',
  sourcesContent: [ '.a {\n\ttext-align: center;\n}\n\n.b {\n\ttext-align: center;\n}\n\n.c {\n\ttext-align: center;\n}\n' ] }

RESULT
 { version: 3,
  sources: [ 'a.css', 'b.css', 'c.css' ],
  names: [],
  mappings: 'AAAA,GACA,kBAAA,ACDA,CDEA,ECDA,kBAAA,ACDA,CDEA,ECDA,kBAAA,CACA',
  file: 'result-postcss.css',
  sourcesContent: 
   [ '.a {\n\ttext-align: center;\n}\n',
     '.b {\n\ttext-align: center;\n}\n',
     '.c {\n\ttext-align: center;\n}\n' ] }
```

**Comment by TrySound** [comment]
Output is correct. But in `dest` need pass double dir.

**Comment by w0rm** [comment]
Closing this issue because gulp-postcss works correctly, mismatch of sourcemap and file.path comes from gulp-concat.

Anyway I can't fix it on my side by changing `from` option of PostCSS to something other than `file.path` because it will introduce regression of this https://github.com/w0rm/gulp-postcss/pull/10

**Comment by TrySound** [comment]
What's happend?

**Comment by w0rm** [comment]
After having discussed it with @floridoo in [gulp-concat issue](https://github.com/wearefractal/gulp-concat/issues/77) I came to conclusion that it is in fact problem of gulp-postcss.

We should use set `from` option of PostCSS to `file.relative` instead of absolute `file.path` to produce correct source map that will be combined with previous steps. But this will break #10 @superlukas

**Comment by w0rm** [comment]
@TrySound I have to use file.relative in postcss to resolve this issue, but it will cause regression. In the meantime you can fix your code with this: 

```js
gulp.task('postcss', function () {
	gulp.src('src/*.css', {base: '.'})
		.pipe(sourcemaps.init())
		.pipe(concat('result-postcss.css'))
		.pipe(postcss([
			require('csswring')
		]))
		.pipe(sourcemaps.write('.'))
		.pipe(gulp.dest('dist'));
});
```

**Comment by TrySound** [comment]
Okey. At least this.

**Comment by floridoo** [comment]
@w0rm couldn't you keep setting `from` to `file.path` and just fix the paths in the source map to be relative before passing it on?

**Comment by w0rm** [comment]
@floridoo I guess I have to patch `sourcemap.result` of existing source map to be relative to `process.cwd()`, and then patch result of postcss source map to equal `file.relative`.

**Comment by floridoo** [comment]
@w0rm Something like this should do the trick (untested):
```diff
// Apply source map to the chain
if (file.sourceMap) {
+  result.map.file = file.relative;
+  result.map.sources = result.map.sources.map(function(source) {
+    return path.relative(file.base, source);
+  });
  applySourceMap(file, result.map.toString())
}
```

**Comment by w0rm** [comment]
@floridoo thanks, will do as you suggested and add a proper test for this.

**Comment by superlukas** [comment]
Looks good to me.

I'm not sure I fully understand how PostCSS works internally. Is there *any* way for a PostCSS plugin to figure out where the file is relative to the CWD unless it's given the absolute path? IIRC (big if) then Gulp has an issue with globs (the top-level directory is missing or something? This might not actually be true) and all PostCSS plugins know about a file is `opts.from` & friends as well as `process.cwd()`, so there would be no way for, say, postcss-import to figure out where a file is relative to the CWD, which breaks everything that behaves like relative imports (relative asset inlining, what have you). I think @MoOx might known more about this.

Basically, what I want to know is if it's possible at all for gulp-postcss to set `opts.from` to anything but an absolute path without breaking plugins down the line that want to work with relative files in some way or another.

**Comment by TrySound** [comment]
Great work! Thanks.
But for know sourcemaps doesn't work when I use postcss twice in the stream. Result is the same.

Like this:
```js
	return gulp.src(bower('**/*.css'))
		.pipe($.sourcemaps.init())
		.pipe($.concat('bundle.css'))
		.pipe($.plumber())
		.pipe($.postcss([
			require('autoprefixer-core')()
		]))
		.pipe(gulp.dest('dist/assets/css'))
		.pipe($.rename({suffix: '.min'}))
		.pipe($.plumber())
		.pipe($.postcss([
			require('csswring')({
				removeAllComments: true
			})
		]))
		.pipe($.sourcemaps.write('.'))
		.pipe(gulp.dest('dist/assets/css'));
```

But when I combine both that works well.

**Comment by w0rm** [comment]
You use gulp rename in between, that renames the file but not the sourceMap.file. Move it just before second gulp.dest

**Comment by TrySound** [comment]
Not before `gulp.dest`, but before `sourcemaps.write`. Cause else there is non-min path of sourcemap in the css.
For know it works well.
Thanks!

**Comment by w0rm** [comment]
@TrySound I opened [this issue](https://github.com/hparra/gulp-rename/issues/37) in gulp-rename but the project seems to be abandoned.

**Comment by andrerpena** [comment]
@TrySound , you rep. was the most helpful thing that I found on the internet about minifying and bundling with PostCSS. If you used Less too it would have been perfect. Thanks for posting this.

**Comment by TrySound** [comment]
@andrerpena You'are welcome.
I don't use Less. PostCSS cover all my requirements.
You can ask a question on gitter or email.

**Comment by andrerpena** [comment]
@TrySound I was able to enable Less based on your solution by adding the `gulp-less` plugin. Pretty straight forward. Thanks again.
